File: backend/src/main/java/com/lightgallery/backend/entity/Subscription.java
package com.declutter.backend.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.time.LocalDateTime;

/**
 * Subscription Entity
 * Stores user subscription information
 */
@Data
@EqualsAndHashCode(callSuper = true)
@TableName("subscriptions")
public class Subscription extends BaseEntity {

    /**
     * Subscription ID (Primary Key)
     */
    @TableId(type = IdType.AUTO)
    private Long id;

    /**
     * User ID (Foreign Key)
     */
    @TableField("user_id")
    private Long userId;

    /**
     * Subscription tier: free, pro, max
     */
    @TableField("tier")
    private String tier;

    /**
     * Billing period: monthly, yearly
     */
    @TableField("billing_period")
    private String billingPeriod;

    /**
     * Status: active, expired, cancelled, pending
     */
    @TableField("status")
    private String status;

    /**
     * Payment method: apple_iap, wechat_pay, alipay
     */
    @TableField("payment_method")
    private String paymentMethod;

    /**
     * Subscription start date
     */
    @TableField("start_date")
    private LocalDateTime startDate;

    /**
     * Subscription expiry date
     */
    @TableField("expiry_date")
    private LocalDateTime expiryDate;

    /**
     * Auto-renewal flag (false: disabled, true: enabled)
     */
    @TableField("auto_renew")
    private Boolean autoRenew;

    /**
     * Product ID from payment platform
     */
    @TableField("product_id")
    private String productId;

    /**
     * Original transaction ID for tracking
     */
    @TableField("original_transaction_id")
    private String originalTransactionId;

    /**
     * Last sync timestamp with backend
     */
    @TableField("last_synced_at")
    private LocalDateTime lastSyncedAt;
}

File: Declutter/Models/Subscription/Subscription.swift
//
//  Subscription.swift
//  Declutter
//
//  Created for user-auth-subscription feature
//

import Foundation

struct Subscription: Codable {
    let id: String
    let userId: String
    var tier: SubscriptionTier
    var billingPeriod: BillingPeriod
    var status: SubscriptionStatus
    var startDate: Date
    var expiryDate: Date
    var autoRenew: Bool
    var paymentMethod: PaymentMethod
    var lastSyncedAt: Date
    
    init(
        id: String = UUID().uuidString,
        userId: String,
        tier: SubscriptionTier,
        billingPeriod: BillingPeriod,
        status: SubscriptionStatus,
        startDate: Date,
        expiryDate: Date,
        autoRenew: Bool = true,
        paymentMethod: PaymentMethod,
        lastSyncedAt: Date = Date()
    ) {
        self.id = id
        self.userId = userId
        self.tier = tier
        self.billingPeriod = billingPeriod
        self.status = status
        self.startDate = startDate
        self.expiryDate = expiryDate
        self.autoRenew = autoRenew
        self.paymentMethod = paymentMethod
        self.lastSyncedAt = lastSyncedAt
    }
    
    var isActive: Bool {
        return status == .active && expiryDate > Date()
    }
    
    var isExpired: Bool {
        return expiryDate <= Date()
    }
    
    var daysRemaining: Int {
        let calendar = Calendar.current
        let components = calendar.dateComponents([.day], from: Date(), to: expiryDate)
        return max(0, components.day ?? 0)
    }
}
